src/rust/ide/span-tree/src/action.rs

```diff
@@ -1,21 +1,334 @@
+        match self.node.kind {
+            node::Kind::Argument{removable:true} |
+            node::Kind::Target  {removable:true} => Some(Box::new(move |root| {
+                let parent_crumb = &self.ast_crumbs[..self.ast_crumbs.len()-1];
+                let ast          = root.get_traversing(parent_crumb)?;
+                let new_ast = if let Some(mut infix) = ast::opr::Chain::try_new(ast) {
+                    match self.node.kind {
+                        node::Kind::Target {..} => {infix.erase_target(); }
+                        _                       => {infix.args.pop();     }
+                    }
+                    infix.into_ast()
+                } else {
+                    let mut prefix = ast::prefix::Chain::new_non_strict(ast);
+                    prefix.args.pop();
+                    prefix.into_ast()
+                };
+                root.set_traversing(parent_crumb,new_ast)
+            })),
+            _ => None
+        }
+    }
 }
 
-impl<'a> SpanTreeActions for node::Ref<'a> {
-    fn can_set          (&self) -> bool { false }
-    fn can_insert_before(&self) -> bool { false }
-    fn can_erase        (&self) -> bool { false }
+
+
+
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use Action::*;
+
+    use wasm_bindgen_test::wasm_bindgen_test;
+    use parser::Parser;
+    use ast::HasRepr;
+    use data::text::Index;
+    use data::text::Span;
+    use std::ops::Range;
+
+    #[wasm_bindgen_test]
+    fn actions_in_span_tree() {
+        #[derive(Debug)]
+        struct Case {
+            expr    : &'static str,
+            span    : Range<usize>,
+            action  : Action,
+            expected: &'static str,
+        }
+
+        impl Case {
+            fn run(&self, parser:&Parser) {
+                let ast        = parser.parse_line(self.expr).unwrap();
+                let tree       = ast.generate_tree().unwrap();
+                let span_begin = Index::new(self.span.start);
+                let span_end   = Index::new(self.span.end);
+                let span       = Span::from_indices(span_begin,span_end);
+                let node       = tree.root_ref().find_by_span(&span);
+                let node       = node.expect(format!("Invalid case {:?}: no node with span {:?}",
+                                                     self,span).as_str());
+                let arg        = Ast::new(ast::Var {name:"foo".to_string()},None);
+                let result = match &self.action {
+                    Set          => node.set(&ast,arg),
+                    Erase        => node.erase(&ast),
+                }.unwrap();
+                let result_repr = result.repr();
+                assert_eq!(result_repr,self.expected,"Wrong answer for case {:?}",self);
+            }
+        }
+
+        let cases:&[Case] = &
+            // Setting
+            [ Case{expr:"a + b"    , span:0..5, action:Set  , expected:"foo"            }
```

Comment:
this should be made a very straightforward macro :P Then a line:

`Case{expr:"a + b"    , span:0..5, action:Set  , expected:"foo"            }`

Could be rewritten in:

`Set 0..5 on [a + b] = [foo]`